File: examples/ikavery/solana/packages/sdk/src/ix/approve.ts
import { PublicKey, SystemProgram, TransactionInstruction } from '@solana/web3.js';

import {
	AUTH_PUBKEY_BYTES,
	AUTH_SIGNATURE_BYTES,
	IKAVERY_PROGRAM_ID,
	IX_APPROVE,
	MAX_CLIENT_DATA_JSON_BYTES,
	SYSVAR_INSTRUCTIONS_ID,
	SYSVAR_RENT_ID,
} from '../constants';
import { approvalPda, memberIdHash } from '../pda';
import { writeBytes, writeU8, writeU16le } from './encode';
import { credentialArgs, type AuthCredential } from './types';

export interface ApproveParams {
	/** Recovery PDA. */
	recovery: PublicKey;
	/** Proposal PDA being approved. */
	proposal: PublicKey;
	/**
	 * Pays rent for the new Approval PDA. Decoupled from the auth identity:
	 * any wallet can sponsor the vote.
	 */
	payer: PublicKey;
	/**
	 * Member-id slot of the approver (the canonical `[scheme, pubkey]` form
	 * — `packMemberSlot` / `packSolanaMember` produce these). Used to
	 * compute the `member_id_hash` UncheckedAccount address that the program
	 * uses as the Approval PDA's per-credential discriminator seed.
	 */
	memberSlot: Uint8Array;
	/** Credential authorising the vote. Must resolve to `memberSlot`. */
	credential: AuthCredential;
}

export interface ApproveIx {
	ix: TransactionInstruction;
	approval: PublicKey;
	memberIdHashAddress: PublicKey;
}

export function buildApproveIx(params: ApproveParams): ApproveIx {
	const memberIdHashAddress = memberIdHash(params.memberSlot);
	const approval = approvalPda(params.proposal, memberIdHashAddress);
	const cred = credentialArgs(params.credential);

	const dataLen = 1 + 1 + AUTH_PUBKEY_BYTES + MAX_CLIENT_DATA_JSON_BYTES + 2 + AUTH_SIGNATURE_BYTES;
	const data = new Uint8Array(dataLen);
	let off = 0;
	off = writeU8(data, off, IX_APPROVE);
	off = writeU8(data, off, cred.authScheme);
	off = writeBytes(data, off, cred.authPubkey, AUTH_PUBKEY_BYTES);
	off = writeBytes(data, off, cred.clientDataJson, MAX_CLIENT_DATA_JSON_BYTES);
	off = writeU16le(data, off, cred.clientDataJsonLen);
	off = writeBytes(data, off, cred.authSignature, AUTH_SIGNATURE_BYTES);

	const ix = new TransactionInstruction({
		programId: IKAVERY_PROGRAM_ID,
		keys: [
			{ pubkey: params.recovery, isSigner: false, isWritable: false },
			{ pubkey: params.proposal, isSigner: false, isWritable: true },
			{ pubkey: memberIdHashAddress, isSigner: false, isWritable: false },
			{ pubkey: approval, isSigner: false, isWritable: true },
			{ pubkey: params.payer, isSigner: true, isWritable: true },
			{
				pubkey: SYSVAR_INSTRUCTIONS_ID,
				isSigner: false,
				isWritable: false,
			},
			{ pubkey: SYSVAR_RENT_ID, isSigner: false, isWritable: false },
			{ pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
		],
		data: Buffer.from(data),
	});

	return { ix, approval, memberIdHashAddress };
}

File: examples/ikavery/solana/packages/program/src/instructions/approve.rs
use crate::auth::challenges;
use crate::auth::members;
use crate::auth::precompile;
use crate::auth::{auth_error_to_program_error, verify_credential};
use crate::error::IkaveryError;
use crate::state::{
    Approval, ApprovalInner, MemberSlot, Proposal, Recovery, AUTH_PUBKEY_BYTES,
    AUTH_SIGNATURE_BYTES, MAX_CLIENT_DATA_JSON_BYTES, STATUS_ACTIVE, STATUS_APPROVED,
};
use quasar_lang::prelude::*;
use solana_address::Address;
use solana_sha256_hasher::hashv;

#[derive(Accounts)]
pub struct Approve {
    pub recovery: Account<Recovery>,
    #[account(mut, has_one(recovery))]
    pub proposal: Account<Proposal>,
    /// Carries the credential's `member_id_hash` via its own account
    /// address. Not loaded; the handler verifies the supplied hash actually
    /// derives from the credential before any state mutates. This is the
    /// only way to feed a per-credential value into the `Approval` PDA's
    /// seed expression — Quasar's `address = …` is evaluated in the
    /// accounts-struct scope and can't reach instruction arguments.
    pub member_id: UncheckedAccount,
    #[account(
        init,
        payer = payer,
        address = Approval::seeds(proposal.address(), member_id.address()),
    )]
    pub approval: Account<Approval>,
    /// Whoever pays for the Approval PDA's rent. Decoupled from the auth
    /// identity: a sponsor wallet can pay while a passkey-derived
    /// credential authorises the vote.
    #[account(mut)]
    pub payer: Signer,
    pub instructions_sysvar: UncheckedAccount,
    pub rent: Sysvar<Rent>,
    pub system_program: Program<SystemProgram>,
}

impl Approve {
    #[inline(always)]
    #[allow(clippy::too_many_arguments)]
    pub fn approve(
        &mut self,
        auth_scheme: u8,
        auth_pubkey: [u8; AUTH_PUBKEY_BYTES],
        client_data_json: [u8; MAX_CLIENT_DATA_JSON_BYTES],
        client_data_json_len: u16,
        auth_signature: [u8; AUTH_SIGNATURE_BYTES],
    ) -> Result<(), ProgramError> {
        let cred_pubkey = members::pubkey_slice(auth_scheme, &auth_pubkey)?;
        let member_slot: MemberSlot = members::credential_slot(auth_scheme, cred_pubkey)?;
        let member_id = members::slot_id(&member_slot)?;
        let _ = members::find_index(self.recovery.members(), member_id)
            .ok_or(IkaveryError::NotAMember)?;

        // The hash carried by `member_id` must match what we'd compute from
        // the credential — otherwise a caller could pin their vote to a
        // different member's slot and dodge double-vote checks.
        let expected_hash = Address::new_from_array(hashv(&[member_id]).to_bytes());
        if &expected_hash != self.member_id.address() {
            return Err(IkaveryError::WrongSigner.into());
        }
        let member_id_hash = expected_hash;

        require!(
            self.proposal.status == STATUS_ACTIVE,
            IkaveryError::ProposalNotActive
        );

        let proposal_index: u32 = self.proposal.proposal_index.into();
        let challenge = challenges::approve(
            self.recovery.recovery_id.as_array().as_slice(),
            proposal_index as u64,
        );

        let cdj_len = client_data_json_len as usize;
        if cdj_len > MAX_CLIENT_DATA_JSON_BYTES {
            return Err(IkaveryError::BadMessageLength.into());
        }
        let cdj = &client_data_json[..cdj_len];

        let sysvar_view = self.instructions_sysvar.to_account_view();
        precompile::check_sysvar_address(sysvar_view.address())
            .map_err(|_| IkaveryError::BadInstructionsSysvar)?;
        let sysvar_data = sysvar_view
            .try_borrow()
            .map_err(|_| IkaveryError::BadInstructionsSysvar)?;
        verify_credential(
            auth_scheme,
            cred_pubkey,
            &challenge,
            self.payer.address(),
            &sysvar_data,
            cdj,
            &auth_signature,
        )
        .map_err(auth_error_to_program_error)?;

        let approval_count: u16 = self.proposal.approval_count.into();
        let new_approvals = approval_count
            .checked_add(1)
            .ok_or(IkaveryError::ArithmeticOverflow)?;

        self.approval.set_inner(ApprovalInner {
            proposal: *self.proposal.address(),
            member_id_hash,
            approved_at_count: new_approvals,
        });

        self.proposal.approval_count = new_approvals.into();

        let threshold: u16 = self.recovery.threshold.into();
        if new_approvals >= threshold {
            self.proposal.status = STATUS_APPROVED;
        }
        Ok(())
    }
}
